src/pipe.js

```diff
@@ -1,34 +1,28 @@
-var compose = require('./compose');
-var reverse = require('./reverse');
+var _pipe = require('./internal/_pipe');
+var curryN = require('./curryN');
+var identity = require('./identity');
+var reduce = require('./reduce');
 
 
 /**
- * Creates a new function that runs each of the functions supplied as parameters in turn,
- * passing the return value of each function invocation to the next function invocation,
- * beginning with whatever arguments were passed to the initial invocation.
- *
- * `pipe` is the mirror version of `compose`. `pipe` is left-associative, which means that
- * each of the functions provided is executed in order from left to right.
+ * Performs left-to-right function composition. The leftmost function may have
+ * any arity; the remaining functions must be unary.
  *
  * In some libraries this function is named `sequence`.
+ *
  * @func
  * @memberOf R
  * @category Function
- * @sig ((a... -> b), (b -> c), ..., (x -> y), (y -> z)) -> (a... -> z)
- * @param {...Function} functions A variable number of functions.
- * @return {Function} A new function which represents the result of calling each of the
- *         input `functions`, passing the result of each function call to the next, from
- *         left to right.
+ * @sig ((a -> b), (b -> c), ..., (y -> z)) -> (a -> b -> c -> ... -> y -> z)
+ * @param {...Function} functions
+ * @return {Function}
  * @example
  *
- *      var triple = function(x) { return x * 3; };
- *      var double = function(x) { return x * 2; };
- *      var square = function(x) { return x * x; };
- *      var squareThenDoubleThenTriple = R.pipe(square, double, triple);
+ *      var f = R.pipe(Math.pow, R.dec);
  *
- *      //≅ triple(double(square(5)))
- *      squareThenDoubleThenTriple(5); //=> 150
+ *      f(3, 4); // 3^4 - 1
  */
 module.exports = function pipe() {
-  return compose.apply(this, reverse(arguments));
+  return curryN(arguments.length === 0 ? 1 : arguments[0].length,
+                reduce(_pipe, {length: 1, value: identity}, arguments).value);
```

Comment:
As I rode my bicycle home last night I tried to reconcile all these things which *almost* fit together:

  - the composition functions originally proposed compose functions of arbitrary arities…
  - these composition functions *do* have an identity element…
  - which means they can be defined as folds with no special case for zero arguments…
  - but these functions don't allow partial application of curried functions…
  - such as `R.pipe(add, R.map)(10, [1, 2, 3])`…
  - because curried functions report their length according to `f(a, b)` rather than `f(a)(b)`…
  - so could we make `R.add` et al. unary functions and have everything?

I was very excited until I realized that `R.pipe(Math.pow, R.subtract)` would then return a binary function rather than a ternary function. I went to bed thinking I had explored another avenue which doesn't offer a solution.

In writing this now, though, I see that this may not be a problem. We can't write `R.pipe(Math.pow, R.subtract)(3, 4, 5)` but we can write `R.pipe(Math.pow, R.subtract)(3, 4)(5)` or `R.pipe(Math.pow, R.subtract)(3)(4)(5)`. So it seems this really *does* work, unless I'm missing something!

Are we happy to report the length of `R.add` et al. as one? We'd be reporting the minimum number of arguments to which a function should be applied rather than the maximum. This is consistent with variadic functions which report their lengths as zero (rather than infinity).